### PR TITLE
Data: Support object as second argument

### PR DESCRIPTION
**Pull request by wjmao88**
jQuery.data supports passing an object as a second argument and sets data for each entry in that object.

This warns the user about each violating non-camelCase key in the object.

This also adds a type check to avoid a current bug where jQuery.camelCase throws an error because a non-string is passed to it.

### CODE REVIEW

File: src/data.js [hunk 1]
```diff
@@ -3,8 +3,26 @@ var oldData = jQuery.data;
 jQuery.data = function( elem, name, value ) {
 	var curData;
 
+	//Name can be an object, and each entry in the object is meant to be set as data
```
Comment (mgol) [inline src/data.js]: There should be a space after `//`.

File: src/data.js [hunk 2]
```diff
@@ -3,8 +3,26 @@ var oldData = jQuery.data;
...
+		for ( var key in name ) {
+			if ( key !== jQuery.camelCase( key ) ) {
+				migrateWarn( "jQuery.data() always sets/gets camelCased names: " + key );
+				curData[ key ] = name[ key ];
+			} else {
+				sameKeys[ key ] = name[ key ];
+			}
+		}
+
+		oldData.call( this, elem, sameKeys );
```
Comment (mgol) [inline src/data.js]: Why not `oldData.apply( this, arguments )` similarly to how what the other configuration is handled? You wouldn't have to record `sameKeys` at all then.

File: test/data.js [hunk 3]
```diff
@@ -37,6 +37,38 @@ test( "jQuery.data() camelCased names", function( assert ) {
 		} );
 	} );
 
+	div = document.createElement( "div" );
```
Comment (mgol) [inline test/data.js]: Could you move all this code to a separate `test`? An example name: "jQuery.data() camelCased names (mass setter)".

**Comment by mgol** [comment]
@wjmao88 Sorry for keeping you wait this long with the review. Will you have time to address my comments? If not, I can finish it for you.

**Comment by mgol** [comment]
This PR should fix issue #198.

File: src/data.js (hunk 2 above)
Comment (wjmao88) [inline src/data.js]: sameKeys is not the same as the second argument, as it only contains keys that are in camelCase and thus does not need to be handled here. 

The other oldData.apply is actually doing the same if it determines the key is already camel case.

**Comment by wjmao88** [comment]
@mgol Yes, I'll be able to find time within a day or so.

File: src/data.js [hunk 4]
```diff
@@ -3,8 +3,26 @@ var oldData = jQuery.data;
...
+
+		return name;
+	}
+
 	// If the name is transformed, look for the un-transformed name in the data object
-	if ( name && name !== jQuery.camelCase( name ) ) {
+	if ( name && typeof name === "string" && name !== jQuery.camelCase( name ) ) {
```
Comment (mgol) [inline src/data.js]: I don't think this change is necessary. We only support documented scenarios, while we want to add support for the object as the second parameter here at this point we know it's not an object so we're fine.

File: src/data.js (hunk 2 above)
Comment (mgol) [inline src/data.js]: Ah, right, I was comparing it to the second part but there's an early return there if the non-camelCased version of the key was found. It's fine then. 👍

File: src/data.js (hunk 4 above)
Comment (wjmao88) [inline src/data.js]: I agree with your reason, which is why I think it should goes straight to jQuery's own .data method, and will throw the same error as when jquery-migrate is not used.

Comment (dmethvin) [inline src/data.js]: @mgol I see @wjmao88 's point here, it's ensuring the behavior (and error location) is identical in both cases. Are you good with that?

**Comment by mgol** [comment]
Yes, his reasoning sounds reasonable!

**Comment by mgol** [comment]
@wjmao88 Landed, thanks!

**Pull request closed without merging**